File: trees/segment_tree_test.cc
#include "segment_tree.hh"

#include <cassert>
#include <iostream>

/** 
 * Usage: 
 * > g++ segment_tree_test.cc -o segment_tree_test -std=c++20
 * > ./segment_tree_test
 */ 

void testBasic() {
    std::vector<int> data {3, 1, 0, 7, 9};
    std::function<int(int, int)> addFn = [](int x, int y) -> int { return x + y; };
    SegmentTree<int> stree_(data, addFn);
    int partialSum = stree_.query(0, 3);  // expect 11
    std::cout << "sum(data[0...3]) = " << partialSum << std::endl;
    assert(partialSum == 11);

    stree_.update(2, 5);
    // data is now {3, 1, 5, 7, 9}
    partialSum = stree_.query(0, 3);  // expect 16
    std::cout << "sum(data[0...3]) = " << partialSum << std::endl;
    assert(partialSum == 16);

    stree_.update(4, 14);
    stree_.update(0, 6);
    stree_.update(0, 7);
    // data is now {7, 1, 5, 7, 14}
    partialSum = stree_.query(0, 4);  // expect 34
    std::cout << "sum(data[0...4]) = " << partialSum << std::endl;
    assert(partialSum == 34);
    std::cout << "testBasic() passes " << std::endl;
}

int main() {
    testBasic();
    return 0;
}

File: trees/segment_tree_test.py
from segment_tree import SegmentTree 

def test_basic():
    data = [1, 4, 7, 11, 20]
    stree = SegmentTree(data, lambda x, y: x+y)
    psum = stree.query(0, 3)  # sum of inclusive range [0, 3]
    assert psum == 23, f"expected 23 but got {psum}"
    stree.update(0, 5)
    stree.update(2, 6) # data is now [5, 4, 6, 11, 20]
    psum = stree.query(0, 3)
    assert psum == 26, f"Expected 26 but got {psum}"
    stree.update(2, 7)
    stree.update(4, 30) # data is now [5, 4, 7, 11, 30]
    psum = stree.query(3, 4)
    assert psum == 41, f"Expected 41 but got {psum}"
    psum = stree.query(0, 4)
    assert psum == 57, f"Expected 57 but got {psum}"
    print("Success")

if __name__=="__main__":
    test_basic()